File: Lectures/03/src/Circle.java
// Compare with the C++ equivalent in circle.hpp & circle.cpp

class Circle {
  private double x, y;
  private double radius;

  public Circle(double r) {
    this(0, 0, r);
  }

  public Circle(double x, double y, double r) {
    if (r <= 0.0) {
      throw new IllegalArgumentException("radius must be > 0");
    }
    this.x = x;
    this.y = y;
    radius = r;
  }

  public double getX() {
    return x;
  }

  public double getY() {
    return y;
  }

  public double getRadius() {
    return radius;
  }

  public boolean containsPoint(double x, double y) {
    double dx = this.x - x;
    double dy = this.y - y;
    return Math.sqrt(dx*dx + dy*dy) <= radius;
  }

  public double area() {
    return Math.PI * radius * radius;
  }

  public double perimeter() {
    return 2.0 * Math.PI * radius;
  }
}

File: Lectures/03/src/circle.cpp
#include <stdexcept>  // <1>
#include "circle.hpp"  // <2>

Circle::Circle(double x, double y, double r):  // <3>
  xorigin(x), yorigin(y), radius(r)  // <4>
{
  if (radius <= 0.0) {
    throw std::invalid_argument("radius must be > 0");  // <5>
  }
}

bool Circle::containsPoint(double x, double y) const  // <6>
{
  double dx = x - xorigin;
  double dy = y - yorigin;
  return sqrt(dx*dx + dy*dy) <= radius;
}
